### DISCUSSION TITLE
raise same event for transition target?

### DISCUSSION THREAD

**Discussion by dwinston** [statelyai/xstate#1847] [Q&A]
I'd like to use `raise()` to raise the current event, but `raise` does not accept a function of `context` and `event` like e.g. the `assign()` action creator.

Modifying the code example for the [Raise Action](https://xstate.js.org/docs/guides/actions.html#raise-action) in the documentation to illustrate my use case:
```js
import { Machine, actions } from 'xstate';
const { raise } = actions;
const raiseActionDemo = Machine({
  id: 'raisedmo',
  initial: 'get_namespace',
  states: {
    get_namespace: {
      on: {
        TOKEN: [{
          target: 'get_name',
          cond: 'is_namespace',
          actions: ['record_namespace']
        }, {
          target: 'get_name',
          actions: [raise((context, event) => event)]
        }]
      }
    },
    get_name: {
      on: {
        TOKEN: {
          target: 'done',
          actions: ['record_name']
        }
      }
    },
    done: {type: 'final'}
  }
});
```

I'd like to be able, in this example, to record names with or without namespaces, where both components are sent as TOKEN events, e.g.
```js
// one sequence of events
[{type: 'TOKEN', data: 'name_without_namespace'}]
// another sequence of events
[{type: 'TOKEN', data: 'namespace'}, {type: 'TOKEN', data: 'name'}]
```
In my case, the "namespace" is recognizably different than a "name", i.e. it is a different type, distinguishable by e.g. an "is_namespace" transition guard.

Have folks encountered use cases where re-raising the current event (or I suppose raising an event that is a function of the current context and/or event) is desired? If so, how did you accomplish it, since `raise` does not take a function as argument?

**Comment by Andarist** [comment]
U can use send with the internal target - im on mobile and i dont recall the exact syntax so please search through our tests or PRs of mine as from what I remember I was the one adding support for this.

@davidkpiano this seems like a weird limitation of SCXML. I believe that we should just support those things in `raise` and always map that to the send action with the internal target. It would simplify the implementation as we wouldnt have to deal with 2 things that do pretty much the same thing. Converting `<raise/>` to XState would still be possible this way. WDYT?

**Reply by ilovett** [reply]
@Andarist @davidkpiano I have the same issue, I would expect `raise` to behave the same way as `send` in terms of taking `(context, event)` callback.  How can I bubble up the exact same event to root level with `send` ?

**Reply by lucasbraganca-tw** [reply]
@ilovett you probably have moved on on this but it can be done like so:
```javascript
actions.send((_, evt) => evt, { to: '#_internal' })
```

It's far from ideal since the action rely on internal/private strings used by xstate and it by no means clearly reveals its intention, but it works. An API like the one @dwinston proposed in his example would certainly be better.

**Comment by huan** [comment]
> support those things in raise and always map that to the send action with the internal target.

+1 :smiley:
